src/main/java/com/microsoft/graph/requests/upload/UploadSliceRequestBuilder.java

```diff
@@ -0,0 +1,82 @@
+import com.microsoft.kiota.*;
+import com.microsoft.kiota.serialization.Parsable;
+import com.microsoft.kiota.serialization.ParsableFactory;
+import okhttp3.Response;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URISyntaxException;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+
+public class UploadSliceRequestBuilder<T extends Parsable> {
+
+    private UploadResponseHandler responseHandler;
+    private RequestAdapter requestAdapter;
+    private String urlTemplate;
+
+    private long rangeBegin;
+    private long rangeEnd;
+
+    private long totalSessionLength;
+
+    private int rangeLength;
+
+    private ParsableFactory<T> factory;
+
+    public UploadSliceRequestBuilder(@Nonnull String sessionUrl, @Nonnull RequestAdapter requestAdapter, @Nonnull ParsableFactory<T> factory, long rangeBegin, long rangeEnd, long totalSessionLength ) {
+
+        this.urlTemplate = Objects.requireNonNull(sessionUrl);
+        this.requestAdapter = Objects.requireNonNull(requestAdapter);
+        this.factory = Objects.requireNonNull(factory);
+        this.rangeBegin = rangeBegin;
+        this.rangeEnd = rangeEnd;
+        this.totalSessionLength = totalSessionLength;
+        this.responseHandler = new UploadResponseHandler(null);
+    }
+
+    public CompletableFuture<UploadResult<T>> putAsync(InputStream stream) {
+        RequestInformation requestInformation = this.createPutRequestInformation(stream);
+        NativeResponseHandler nativeResponseHandler = new NativeResponseHandler();
+        requestInformation.setResponseHandler(nativeResponseHandler);
+        return this.requestAdapter.sendPrimitiveAsync(requestInformation, InputStream.class, null)
+            .thenCompose(i -> {
+               try {
+                   return (CompletableFuture<UploadResult<T>>)responseHandler.handleResponse((Response) nativeResponseHandler.getValue(), factory);
+               } catch (IOException ex) {
+                   return new CompletableFuture<UploadResult<T>>() {{
+                       this.completeExceptionally(ex);
+                   }};
+               } catch (URISyntaxException ex) {
+                   return new CompletableFuture<UploadResult<T>>() {{
+                       this.completeExceptionally(ex);
+                   }};
+               }
+            });
+    }
+
+    public RequestInformation createPutRequestInformation(InputStream stream) {
+        Objects.requireNonNull(stream);
+        RequestInformation  requestInfo = new RequestInformation() {{
+            httpMethod = HttpMethod.PUT;
+            urlTemplate = this.urlTemplate;
+        }};
+        requestInfo.setStreamContent(stream);
+        requestInfo.headers.add("Content-Range", "bytes " + String.format("bytes %f-%f/%f", this.rangeBegin, this.rangeEnd, this.totalSessionLength));
+        requestInfo.headers.add("Content-Length", ""+this.rangeLength);
```

Comment:
Ah, you're right. Will add a line to set the rangeLength upon creation so that this does not occur. At the moment I am keeping the variables such as rangeBegin, rangeEnd , rangeLength etc private. I could make them public and thus remove the need for some get methods since we could then access them directly. Wondering if there may be a negative side affect to making these public though?

Reply:
Maybe its best to keep them private and let them be initialized through the constructor for now. If a need arises we can change the visibility without breaking anything.